### ISSUE TITLE
[RFC] Support MemoryView for Tensors

### ISSUE THREAD

**Issue by haocizhang** [pytorch/pytorch#69491]
## 🚀 Feature
[RFC] Support MemoryView for Tensors

## Motivation
When we are checkpointing the state dict to remote storage, we currently move entire tensors from GPU to CPU before uploading the bytes which is pretty memory intensive. Instead of moving entire tensors to CPU before uploading it to remote storage, we are considering to do this via streaming — taking a tensor from GPU to CPU chunk by chunk and prevent copying the entire tensor at once. On the checkpoint loading path, we hope to do something similar and prevent caching the entire tensor in CPU before moving the bytes to GPU.

An approach to do this is to use the Tensor's underlying Storage buffer and pass an offset and length for that buffer to a library responsible for writing out bytes over the wire or to disk which avoids making a copy of the data. However, we were not able to find something similar that is possible at the Python level without avoiding the numpy dependency. This also has been brought up previously [here](https://github.com/pytorch/pytorch/issues/19143).

## Pitch

We are proposing a MemoryView class for tensors which is similar to the memoryview available in python. The couple of functionalities it supports are:

* From tensor to MemoryView:
    * For a tensor in CPU or GPU memory, given (tensor, offset, size), return the memory view so that the caller can get a view of the bytes. Users can only move MemoryView across hosts and devices (in the above example, users will construct a memory view for the tensors in GPU, move it to CPU and stream out bytes to storage).
* From MemoryView to tensor:
    * Assuming that the source memory view and the destination tensor are residing on the same device (GPU/DRAM). Given a memory view and a tensor, populate the tensor using the bytes in the memory view. Users can later on move the tensor across host/device or fill a part of another tensor using this tensor.

**Note that we are not planning to support non-contiguous tensors in the first version.**

## Interfaces

1) We will expose the torch.MemoryView class which is similar to Python’s memory_view. The init function takes in either a tensor or a python memory view. 
```
class MemoryView:
    def __init__(
        self, 
        input_view: Optional[memory_view], 
        input_tensor: Optional[Tensor],
        input_tensor_offset: Optional[int],
        memory_view_size: Optional[int],
    ) -> None:
```
The init function caches the python memory view if a memory view is passed in, otherwise it will convert the tensor into a memory view and cache the result. 

2) We expose a variable for users to understand if the memory view is on cuda or in DRAM.
```
memory_view.is_cuda: bool
```
3) We expose an API for users to move memory view to device from host.
```
def to_device(self, device) -> MemoryView
```
4) Returns a copy of this object in CPU memory. If this object is already in CPU memory and on the correct device, then no copy is performed and the original object is returned.
```
def cpu(self) -> MemoryView
```
5) Fills the content of the tensor using bytes in the memory view.
```
def fill_tensor(self, tensor) -> None
```
6) Returns the length of the memory view
```
def nbytes(self) -> int
```
7) Returns the reference to the Python memory view and enable users to interact with the memory view object itself. (This is designed to work for CPU memory view only as of now because Python’s memoryview has limited support for GPU)
```
def get_py_memoryview(self) -> memoryview
```
## Additional Context / Implementation Details

**Python Layer:** In the Python layer, we will mainly implement the MemoryView class which will interact with C++ APIs. If the user passed in a memory view to the init function, we will cache the view along with its metadata (device, nbytes etc.) If users passed in a tensor instead of a memory view, we will convert it to a python memory view and cache it in the MemoryView class along with its metadata (device, nbytes etc.). The utility functions can be implemented by simply returning the cached metadata in the class (e.g. nbytes, is_cuda, get_py_memoryview). The cpu() operation can be implemented by allocating some bytes in cpu and copying the bytes over, while the to_device() operation can be implemented by converting the memory view to a CPU based tensor, moving the tensor to GPU and finally constructing a MemoryView object using the tensor on device. The initialization function and fill_tensor requires us to implement a couple of C++ interfaces described in the following section.

**C++ Layer:** In the C++ layer, we will need two functions to convert between memory views and tensors:
```
m.def("to_py_memory_view", [](const at::Tensor& tensor, int64_t offset, int64_t size) -> py::memoryview {
        return py::memoryview::from_memory(
            (void*)tensor.data() + (offset + tensor.storage_offset()) * tensor.element_size(), 
            size, 
            true));
      });

m.def("from_py_memory_view", [](
    const at::Tensor& tensor, const py::memoryview& memoryview) -> void {
        populate(tensor, memoryview);
      });
```

**Comment by pritamdamania87** [comment]
@gchanan @mruberry Was wondering if we can tag relevant folks from the core team to review this proposal :)

**Comment by xunnanxu** [comment]
Could you shed some light on:
1. How do we plan to manage/expose the memory layout? and how does that work together with `fill_tensor()`? Does the layout need to be the same as the target tensor implicitly?
2. While there's no plan to support non-contiguous tensor in the first cut, wonder how this is extensible to do so?
3. Does this imply python implementation? I assume the answer is no. In that case could you explain how C++/Python would fill data into the `MemoryView` before `fill_tensor()` is called?
4. How potentially is `MemoryView` in this proposal better than `Union[memory_view, Tensor]` at high level other than the uniform api?

**Comment by mruberry** [comment]
cc @rgommers who may be able to provide information on NumPy's `memoryview` support

I'm not a fan of creating a customer `MemoryView` class in PyTorch. This functionality is interesting but it can be built using existing PyTorch components if desired. I think we would accept an implementation of the buffer protocol for PyTorch CPU tensors to improve consistency with NumPy and support the scenarios discussed in https://github.com/pytorch/pytorch/issues/19143. I'm not sure there's a good way to implement the buffer protocol for PyTorch tensors on other devices, however.

**Comment by rgommers** [comment]
> cc @rgommers who may be able to provide information on NumPy's `memoryview` support

`memoryview` doesn't require any special handling beyond supporting the Python buffer protocol ([docs](https://docs.python.org/3/c-api/buffer.html), [PEP 3118](https://www.python.org/dev/peps/pep-3118/)).

Potential issues with implementing the buffer protocol:
- It will change what `np.asarray(a_tensor)` does to use the buffer protocol instead of `Tensor.__array__`. So the checks that function does, like raising exceptions when a tensor needs detaching first or isn't on CPU, need to be reimplemented.
- `__array__` handles `__torch_function__` [like so](https://github.com/pytorch/pytorch/blob/master/torch/_tensor.py#L717), the buffer protocol is a C/C++-only implementation so needs to be able to deal with that separately.
- The related `__cuda_array_interface__` gives a clear error message and works with `hasattr` (https://github.com/pytorch/pytorch/blob/master/torch/_tensor.py#L764-L766), and is dynamically hidden in `__dir__` when not available (https://github.com/pytorch/pytorch/blob/master/torch/_tensor.py#L708-L710). This isn't (easily) possible for the buffer protocol.
- This comment may be relevant (it was about `__array_interface__` support, but the same applies to the buffer protocol) https://github.com/pytorch/pytorch/issues/54138#issuecomment-803363164
- And this one too https://github.com/pytorch/pytorch/issues/58036#issuecomment-839656590

Perhaps more importantly:
- the buffer protocol is already supported in `torch.frombuffer`, see gh-59077.
- The RFC does not consider using DLPack, which is very similar to the buffer protocol but works on both CPU and GPU. I think it should meet the requirements of the motivation for this RFC. See [torch.utils.dlpack.from_dlpack](https://pytorch.org/docs/stable/dlpack.html?highlight=from_dlpack#torch.utils.dlpack.from_dlpack). Ignore the warning in the docs, it doesn't apply to `from_dlpack(a_tensor)` (I'll go and fix that). If `from_dlpack` indeed works for this use case, I think we should move it to the top-level `torch` namespace to make it more discoverable - this kind question is coming up too often.

> I'm not a fan of creating a custom `MemoryView` class in PyTorch

Agreed, this is not a good idea.

**Comment by rgommers** [comment]
> However, we were not able to find something similar that is possible at the Python level without avoiding the numpy dependency

So what is supported by `torch.Tensor` is:
- `__array__` (requires `numpy` dependency, CPU-only)
- `__cuda_array_interface__` (CUDA-only)
- `__dlpack__` (multi-device support, for now CPU/CUDA/ROCm, but can be extended)

**Comment by vadimkantorov** [comment]
Yeah, in https://github.com/pytorch/pytorch/issues/58036#issuecomment-839656590 I mainly argue for `torch.as_tensor` to be able to consume various tensor-like objects: torch tensors (not sure what to do with tensor subclasses, but at least regular tensors should be supported), numpy arrays, PIL images, HDF5 datasets, DLPack capsules. Normally, it just means that torch.as_tensor should consume __array_interface__... It sometimes is very useful to have a consume-all generic point similar to numpy.asarray
